test/Runtime/type_discovery.swift

```diff
@@ -0,0 +1,27 @@
+// RUN: %target-run-simple-swift
+// REQUIRES: executable_test
+// XFAIL: OS=windows-msvc
+
+// Disabled on Windows because __ImageBase is not linked properly when building
+// the test target, so the use of #dsohandle causes an error.
+
+import StdlibUnittest
+
+protocol P { }
+struct A: P { }
+class B: P { }
+enum C: P { }
+struct D /* does not conform */ { }
+
+// Repeat the test a few times to exercise the imageAddress cache.
+for _ in 0 ..< 10 {
+    var typeList = [P.Type]()
+
+    _enumerateTypes(fromImageAt: #dsohandle, conformingTo: P.self) { type, _ in
```

Comment:
If you can think of a way to write that test please let me know. As far as I can see, because the order in which types are enumerated is unspecified, there's no good way to test that stopping stopped at the right point—we might stop at any point in the enumeration including on the first iteration or the last.

Reply:
I'm imagining a simple test that just stops after one iteration, e.g. "find ANY type conforming to this protocol"

Reply:
Something like:

```
var iterations = 0

_enumerateTypes(conformingTo: P.self) { _, stop in
    iterations += 1
    stop = true
}

assert(iterations == 1)
```

Reply:
That could work, yeah.